AgentCredential.from_dict accepts any token_type from a persisted record

`agent_identity.py:329` deserializes a credential's `token_type` with `str(d.get("token_type", "opaque"))`, while the field is declared `Literal["opaque", "jwt"]` at `:287`. The widening is invisible at runtime, and mypy names it:

```
src/bernstein/core/agents/agent_identity.py:329: error: Argument "token_type" to
"AgentCredential" has incompatible type "str"; expected "Literal['opaque', 'jwt']"
```

Reproduces on clean `main`. Nothing is red about it because `uv run mypy src` is advisory in `ci.yml`, and the strict zone covers `core/identity/`, not `core/agents/`.

## Why this is more than a typing nit

`from_dict` is the boundary where a record on disk becomes an object the rest of the code trusts. Every other field there is coerced **and** constrained — `_credential_tenant_id`, `_string_list` — but `token_type` is only coerced.

So a record carrying `"token_type": "anything"` loads without complaint, and then flows into decisions like `if cred.token_type != "jwt"` (`:825`), where an unrecognised value silently takes a branch instead of being refused. The dangerous direction is narrow today. The shape is still wrong: an unknown enum value belongs refused at the boundary, not routed by whichever comparison happens to run first.

## What to build

Validate at deserialization, in the same style as the helpers beside it: an unknown `token_type` does not produce a loaded credential.

## What proves it

Tests named for the property they protect:

1. a record whose `token_type` is outside the literal is refused, and the message names the offending value
2. both valid values round-trip through `to_dict` / `from_dict` unchanged
3. an absent `token_type` still defaults to `opaque` — records written before this change must stay loadable

And `uv run mypy src/bernstein/core/agents/agent_identity.py` reports 0 errors.

## The decision left to whoever takes it

Whether "refused" means raising, or degrading to `opaque` with a recorded warning.

Raising is honest but can make a corrupted store unreadable at a bad moment. Degrading keeps the store loadable but silently rewrites a security-relevant field, which is its own way to be wrong. Both are defensible; state which you picked and why in the PR.

## Out of scope

The other fields in `from_dict`. The JWT minting path at `:679`. The merge-gate work in #3914, which touches this file for unrelated reasons.

Found while reviewing #4002.

--- a/src/bernstein/core/agents/agent_identity.py
+++ b/src/bernstein/core/agents/agent_identity.py
@@ -301,6 +301,16 @@ def _credential_token_type(raw: Any) -> Literal["opaque", "jwt"]:
     would rewrite a security-relevant field on the way in and lose the fact
     that the store holds something nothing wrote.
 
+    The ``isinstance`` check is what makes that refusal reachable for every
+    stored shape rather than most of them.  A membership test hashes its left
+    operand, and JSON persists two values that cannot be hashed: a list and
+    an object.  Without the guard those two raise ``TypeError: unhashable
+    type: 'list'`` out of the lookup itself - caught by
+    :meth:`AgentIdentityStore._read_identity` like any other refusal, so the
+    store stays readable, but naming neither the field nor the value, which
+    is the whole point of the message below.  It is also the shape
+    :func:`_credential_tenant_id` beside it already uses.
+
     Args:
         raw: The stored value, or ``"opaque"`` when the record has no
             ``token_type`` key.
@@ -309,7 +319,7 @@ def _credential_token_type(raw: Any) -> Literal["opaque", "jwt"]:
         ValueError: The record carries a ``token_type`` outside
             ``Literal["opaque", "jwt"]``.
     """
-    if raw in _CREDENTIAL_TOKEN_TYPES:
+    if isinstance(raw, str) and raw in _CREDENTIAL_TOKEN_TYPES:
         return cast('Literal["opaque", "jwt"]', raw)
     raise ValueError(f"credential token_type must be one of {sorted(_CREDENTIAL_TOKEN_TYPES)}, got {raw!r}")
 

--- a/src/bernstein/core/agents/agent_identity.py
+++ b/src/bernstein/core/agents/agent_identity.py
@@ -153,6 +153,11 @@ def permissions_for_role(role: str) -> frozenset[str]:
 # former is a pre-field record; see :func:`_credential_tenant_id`.
 _TENANT_KEY_ABSENT: Final[object] = object()
 
+# The two kinds a credential may declare, kept beside the ``token_type``
+# annotation on :class:`AgentCredential` so the deserialiser and the type
+# cannot drift apart; see :func:`_credential_token_type`.
+_CREDENTIAL_TOKEN_TYPES: Final[frozenset[str]] = frozenset({"opaque", "jwt"})
+
 
 def _string_list(raw: Any, field: str) -> list[str]:
     """Return a persisted list-of-strings field, refusing any other shape.
@@ -272,6 +277,43 @@ def _credential_tenant_id(raw: Any) -> str:
     return normalize_tenant_id(raw)
 
 
+def _credential_token_type(raw: Any) -> Literal["opaque", "jwt"]:
+    """Return the token kind a persisted credential declares.
+
+    ``token_type`` selects which validation a token gets: ``_validate_jwt_claims``
+    refuses anything whose credential does not say ``"jwt"``, so authentication
+    falls through to the opaque hash comparison for every other value.  A
+    ``str()`` coercion of the stored value therefore does not merely widen a
+    type - it routes an authentication decision on a value nothing has
+    established, and the routing is by whichever comparison runs first rather
+    than by a recognised kind.
+
+    An unknown kind is refused here instead, in the same style as
+    :func:`_credential_tenant_id` beside it: the boundary where a record on
+    disk becomes an object the rest of the code trusts is where a value that
+    is not one of the two real kinds stops.
+
+    Leniency is for age, not for content: a record written before the field
+    existed carries no ``token_type`` at all and is an opaque credential,
+    which is what the dataclass default has always said.  A record that
+    carries the key with something else in it asserted a kind, and an
+    unrecognised assertion is refused rather than defaulted - defaulting it
+    would rewrite a security-relevant field on the way in and lose the fact
+    that the store holds something nothing wrote.
+
+    Args:
+        raw: The stored value, or ``"opaque"`` when the record has no
+            ``token_type`` key.
+
+    Raises:
+        ValueError: The record carries a ``token_type`` outside
+            ``Literal["opaque", "jwt"]``.
+    """
+    if raw in _CREDENTIAL_TOKEN_TYPES:
+        return cast('Literal["opaque", "jwt"]', raw)
+    raise ValueError(f"credential token_type must be one of {sorted(_CREDENTIAL_TOKEN_TYPES)}, got {raw!r}")
+
+
 @dataclass
 class AgentCredential:
     """Bearer token for agent-to-server authentication.
@@ -326,7 +368,7 @@ class AgentCredential:
             created_at=float(d.get("created_at", 0)),
             expires_at=float(d.get("expires_at", 0)),
             revoked=bool(d.get("revoked", False)),
-            token_type=str(d.get("token_type", "opaque")),
+            token_type=_credential_token_type(d.get("token_type", "opaque")),
             algorithm=str(d.get("algorithm", "HS256")),
             jti=str(d.get("jti", "")),
             tenant_id=_credential_tenant_id(d.get("tenant_id", _TENANT_KEY_ABSENT)),

--- a/tests/unit/test_agent_identity.py
+++ b/tests/unit/test_agent_identity.py
@@ -466,12 +466,20 @@ class TestAgentCredentialTokenTypeDeserialization:
         assert credential.token_type == stored
         assert AgentCredential.from_dict(credential.to_dict()).token_type == stored
 
-    @pytest.mark.parametrize("stored", ["anything", "JWT", "opaque ", "", None, 1, True])
+    @pytest.mark.parametrize("stored", ["anything", "JWT", "opaque ", "", None, 1, True, [], {}])
     def test_unknown_token_type_is_refused_and_named(self, stored: object) -> None:
         """The message names the offending value, so the record is findable.
 
         A store holding one bad record is repaired by locating it; a refusal
         that only names the field leaves every credential a suspect.
+
+        The last two cases are the reason the membership test is guarded by
+        ``isinstance``: a list and an object are the only two values JSON can
+        persist that a set lookup cannot hash, so without the guard they
+        raise ``TypeError: unhashable type`` from the lookup rather than the
+        named ``ValueError``.  Both are here rather than one, because they
+        are two different unhashable shapes and a guard covering one is not
+        evidence about the other.
         """
         with pytest.raises(ValueError, match="token_type") as excinfo:
             AgentCredential.from_dict({"token_hash": "abc", "token_type": stored})

--- a/tests/unit/test_agent_identity.py
+++ b/tests/unit/test_agent_identity.py
@@ -443,6 +443,63 @@ class TestAgentCredentialTenantDeserialization:
         assert "session-corrupt" not in listed
 
 
+class TestAgentCredentialTokenTypeDeserialization:
+    """The persisted ``token_type`` selects which validation a token gets.
+
+    ``_validate_jwt_claims`` refuses any credential that does not say
+    ``"jwt"``, so an unrecognised kind is not inert - it routes the token to
+    the opaque hash comparison instead of being refused.  ``from_dict``
+    establishes the value as one of the two real kinds rather than coercing
+    whatever was on disk into one that some comparison will happen to accept.
+    """
+
+    def test_absent_token_type_resolves_to_opaque(self) -> None:
+        """Records written before the field existed still load."""
+        credential = AgentCredential.from_dict({"token_hash": "abc"})
+
+        assert credential.token_type == "opaque"
+
+    @pytest.mark.parametrize("stored", ["opaque", "jwt"])
+    def test_both_kinds_round_trip_unchanged(self, stored: str) -> None:
+        credential = AgentCredential.from_dict({"token_hash": "abc", "token_type": stored})
+
+        assert credential.token_type == stored
+        assert AgentCredential.from_dict(credential.to_dict()).token_type == stored
+
+    @pytest.mark.parametrize("stored", ["anything", "JWT", "opaque ", "", None, 1, True])
+    def test_unknown_token_type_is_refused_and_named(self, stored: object) -> None:
+        """The message names the offending value, so the record is findable.
+
+        A store holding one bad record is repaired by locating it; a refusal
+        that only names the field leaves every credential a suspect.
+        """
+        with pytest.raises(ValueError, match="token_type") as excinfo:
+            AgentCredential.from_dict({"token_hash": "abc", "token_type": stored})
+
+        assert repr(stored) in str(excinfo.value)
+
+    def test_unknown_token_type_does_not_authenticate(self, tmp_path: Path) -> None:
+        """The refusal reaches authentication as a miss, not as a 500.
+
+        Before this change the record loaded, ``token_type != "jwt"`` sent it
+        down the opaque branch, and the identity authenticated on its token
+        hash alone.  The refusal now travels the same path an unusable
+        ``tenant_id`` already does - through ``_load``, which skips the file.
+        """
+        import json
+
+        store = AgentIdentityStore(tmp_path)
+        identity, token = store.create_identity("session-jwt", "backend")
+        path = tmp_path / "agent_identities" / f"{identity.id}.json"
+        payload = json.loads(path.read_text())
+        assert payload["credential"]["token_type"] == "jwt"
+        payload["credential"]["token_type"] = "anything"
+        path.write_text(json.dumps(payload))
+
+        assert store.authenticate(token) is None
+        assert [found.id for found in store.list_identities()] == []
+
+
 class TestCorruptIdentityDoesNotBreakAuthentication:
     """A record that will not deserialise authenticates as nobody, not as a 500.
 

